server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregationBuilder.java

```diff
@@ -337,15 +337,21 @@ public ZoneId timeZone() {
     protected final ValuesSourceAggregatorFactory doBuild(QueryShardContext queryShardContext, AggregatorFactory parent,
                                                           Builder subFactoriesBuilder) throws IOException {
         ValuesSourceConfig config = resolveConfig(queryShardContext);
-        if (queryShardContext.getValuesSourceRegistry().isRegistered(getType())) {
-            // Only test if the values source type is valid if the aggregation uses the registry
-            AggregatorSupplier supplier = queryShardContext.getValuesSourceRegistry().getAggregator(config, getType());
+        if (queryShardContext.getValuesSourceRegistry().isRegistered(getRegistryKey())) {
+            /*
+            if the aggregation uses the values source registry, test if the resolved values source type is compatible with this aggregation.
+            This call will throw if the mapping isn't registered, which is what we want.  Note that we need to throw from here because
+            AbstractAggregationBuilder#build, which called this, will attempt to register the agg usage next, and if the usage is invalid
+            that will fail with a weird error.
+             */
+            queryShardContext.getValuesSourceRegistry().getAggregator(getRegistryKey(), config);
```

Comment:
I wonder if the factory should be registering the usage. Or if the builder should resolve the registry entry first and pass it down to the factory inside `innerBuild`.....

Reply:
Honestly, I'm not really sure why we don't build the aggregator directly here.  We have all the shard specific information at this point, and we've already parsed everything.  But it would be a huge change to do that, and I'm not ready to take it on today (or any time soon).  But longer term, I think it would make sense for the builder to just create an aggregator directly and not muck about with the factory layer.  In that spirit, I don't want to push more responsibility into the factory.
Also, this is a 1000+ line delta, I don't want to touch anything I don't strictly have to.